src/main/scala/iog/psg/cardano/CardanoApiCodec.scala

```diff
@@ -418,21 +432,25 @@ object CardanoApiCodec {
       }
     }
 
-    def toWallet: Future[CardanoApiResponse[Wallet]]
-    = to[Wallet](Unmarshal(_).to[CardanoApiResponse[Wallet]])
-
-    def toWallets: Future[CardanoApiResponse[Seq[Wallet]]]
-    = to[Seq[Wallet]](Unmarshal(_).to[CardanoApiResponse[Seq[Wallet]]])
-
-
-    def toWalletAddressIds: Future[CardanoApiResponse[Seq[WalletAddressId]]]
-    = to[Seq[WalletAddressId]](Unmarshal(_).to[CardanoApiResponse[Seq[WalletAddressId]]])
-
-    def toFundPaymentsResponse: Future[CardanoApiResponse[FundPaymentsResponse]]
-    = to[FundPaymentsResponse](Unmarshal(_).to[CardanoApiResponse[FundPaymentsResponse]])
+    final def decodeInStream[T](response: HttpResponse, jsonPath: String)(implicit um: Unmarshaller[String, T]): Future[CardanoApiResponse[Seq[T]]] =
+      decodeResponseEntityOrHandleError(response, () =>
+        response.entity.dataBytes
+          .via(JsonReader.select(jsonPath))
+          .mapAsync(4)(bs => unmarshalOrRecoverToUnparseable[T](bs.utf8String))
+          .runWith(Sink.seq).map(flattenCardanoApiResponses)
+      )
+
+    private def unmarshalOrRecoverToUnparseable[T](utf8String: String)(implicit um: Unmarshaller[String, T]): Future[CardanoApiResponse[T]] =
+      Unmarshal(utf8String).to[T].map(Right(_)).recover {
+        case e: Exception => errorUnparseableResult(e)
+      }
 
-    def toCreateTransactionResponses: Future[CardanoApiResponse[Seq[CreateTransactionResponse]]]
-    = to[Seq[CreateTransactionResponse]](Unmarshal(_).to[CardanoApiResponse[Seq[CreateTransactionResponse]]])
+    private def flattenCardanoApiResponses[T](resp: Seq[CardanoApiResponse[T]]): CardanoApiResponse[Seq[T]] =
```

Comment:
This is elegant but it does go through the list twice in the optimistic case. Use a foldLeft with a CardanoResponse[Seq[T]] accumulator. Also this is a 'sequencer' right? Like Future.sequence - probably should be called `sequenceCardanoApiReponses`.

Reply:
hmm then I would propose:

   ```
private def sequenceCardanoApiReponses[T](resp: Seq[CardanoApiResponse[T]]): CardanoApiResponse[Seq[T]] =
      resp.find(_.isLeft) match {
        case Some(Left(error)) => Left(error)
        case None => Right(resp.flatMap(_.toOption))
      }
```

find will look for first Left, faster then foldLeft and current solution

Reply:
It's got the same problem, in the case where there are no errors - which is 'all' the time - it goes through the whole list twice.

Reply:
yes, sure I will do it in a way that will go just 1 time

Reply:
@mcsherrylabs pls check now, not it will work as fast as possible, will break on error and if no error will just go through the list once